advanced_source/cpp_dataset.rst

```diff
@@ -0,0 +1,405 @@
+Custom Dataset and Dataloader in PyTorch C++
+============================================
+
+One of the most powerful components of PyTorch are the dataset utilities which
+help users optimize and parallelize reading and serving data into models.  
+
+In this tutorial, we are going to create a custom dataset that contains 2
+inputs and a label.  This dataset will be inspired by the `MNIST dataset
+<https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/include/torch/data/datasets/mnist.h>`_.
+These datasets can be used for applications that require more than a single
+input and single label.
+
+.. tip::
+
+  We recommend reading the `C++ Frontend Tutorial <cpp_frontend.html>`_ if 
+  you have not done so already.  There you will learn how to setup a C++ PyTorch 
+  project.
+
+Motivation
+----------
+
+We want to leverage the raw speed of C++ for reading data and take advantage of
+C++'s native multithreading with a custom dataset.  Additionally, the ability to
+customize the outputs of our dataset allows us to easily use a wider range of
+PyTorch models.  For example, the current batch of transformer models have
+multiple inputs for the various masks, token type ids, and position ids that are
+used as inputs into the network.  The base PyTorch C++ dataset only supports two
+different tensors ("data" and "target") for each example.  Here we will create a
+dataset that supports 3 different tensors, and we will create the necessary
+files to use the powerful PyTorch dataloader utilities.
+
+Setting Up the Project
+----------------------
+
+As in the frontend tutorial above, we will be using ``cmake`` to compile our
+project.
+
+.. code-block:: cmake
+  :caption: CMakeLists.txt
+  :name: cmakelists-txt
+
+  cmake_minimum_required(VERSION 3.0 FATAL_ERROR)
+
+  # allows us to use Torch_ROOT environmental variable
+  cmake_policy(SET CMP0074 NEW)
+
+  # alternative to set_property(...)
+  set(CMAKE_CXX_STANDARD 14)
+  set(CMAKE_CXX_STANDARD_REQUIRED ON)
+
+  project(custom-dataset 
+          VERSION 0.0.1 
+          DESCRIPTION "Custom Dataset with PyTorch's C++ Frontend")
+
+  # add torch
+  find_package(Torch REQUIRED)
+
+  # put all sources files into a single variable
+  set(SOURCES
+      custom_dataset.cpp
+      main.cpp
+     )
+
+  add_executable(custom-dataset ${SOURCES})
+  target_link_libraries(custom-dataset "${TORCH_LIBRARIES}")
+
+.. note::
+
+  If you need ``CMAKE_PREFIX_PATH`` to be something other than the 
+  location to the ``libtorch`` directory, you can use the environmental 
+  variable ``Torch_DIR`` as an alternative.  To do so, you export 
+  it in your shell as follows:
+
+.. code-block:: shell
+
+  export Torch_DIR=/full/path/to/libtorch/share/cmake/Torch
+  # or
+  export Torch_DIR=$(realpath rel/path/to/libtorch/share/cmake/Torch)
```

Comment:
`rel` means relative path right? if so, can we use full word to make this clear?